### PR TITLE
Fix #34: Assigns the corresponding project owner for the first review of a PR and Fix #19: Add check to ensure all PRs have changelog labels

### LINKED ISSUE #19: Add checks to ensure that a "changelog" label is applied on every PR
We would like any PR that is merged into develop to have a "changelog: -type-" label. This is to help organize the changelog easily during releases. 

PRs must contain exactly one changelog label, and the bot auto-close the PR if no label is found. It will leave a top-level comment asking the PR author to add a changelog label before reopening the PR.

### LINKED ISSUE #34: Assign the corresponding project owner for the first review of a PR once a project label is added
The bot should assign the corresponding project owner once it finds a project label on a PR. The bot should also leave a message saying “Assigning to @XXX for the first-pass review.”
If no project label is found, the bot will close the PR and ask the author to add a project label to the PR before re-opening it. In this way, all open PRs will always follow the review cycle.

### PR DESCRIPTION
**Pull request by apb7**
## Explanation

Fixes #34: Assigns the corresponding project owner for the first review of a PR once a project label is found. If no project label is found, the bot will close the PR with a message to add a project label to the author.

Fixes #19: Adds check to ensure that all PRs have changelog labels. If a PR does not have a changelog label, close the PR with a comment addressed to the author to add the label.

How the checks will function:

screenshot-lh3 googleusercontent com-2019 07 27-02-03-38

~Please note that the project labels need to be modified as discussed (@DubeySandeep is working on it) before this check is deployed.~ (Sandeep has done this. Thanks, Sandeep!)

## Checklist
- [X] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [X] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

File: lib/checkPullRequestLabels.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  var userName = pullRequest.user.login;
+  var projectOwnerLabel;
+  var hasProjectOwnerLabel = labels.some(function(label) {
+    projectOwnerLabel = label.name;
+    return projectOwnerLabel.startsWith('PROJECT');
+  });
+
+  // If the PR has a project owner label, assign the project owner
+  // for the first-pass review of the PR and leave a top-level comment.
+  if(hasProjectOwnerLabel === true) {
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: `if (hasProjectOwnerLabel)`

File: lib/checkPullRequestLabels.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    var labelSubstrings = projectOwnerLabel.split('@');
+    var projectOwner = labelSubstrings[labelSubstrings.length-1].trim();
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: Put spaces around '-'

File: lib/checkPullRequestLabels.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    return;
+  }
+
+  // If no project owner label is found, close the PR.
+  // Ask the author to add the label and re-open it.
+  var stateParams = context.issue({state: 'closed'});
+  await context.github.issues.edit(stateParams);
+
+  var stateCommentParams = context.issue({body: 'Hi, @' + userName +
+    '. This pull request has been closed since it does have a project ' +
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: I looked at the list of labels and as a new contributor, I have no idea what a "project owner label" is. You might want to make this instruction more explicit.

That said ... are you sure it should be "project owner" rather than "project"? I'm not sure how a new contributor is going to know who the project owner is.

Comment (seanlip) [inline lib/checkPullRequestLabels.js]: Also, not entirely sure whether closing is too drastic. But I think that's a question for you + Sandeep + dev workflow team; I'm happy to defer to the joint decision. Maybe set up a quick group chat and ask.

File: lib/checkPullRequestLabels.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  var projectOwnerLabel;
+  var hasProjectOwnerLabel = labels.some(function(label) {
+    projectOwnerLabel = label.name;
+    return projectOwnerLabel.startsWith('PROJECT');
+  });
+
+  // If the PR has a project owner label, assign the project owner
+  // for the first-pass review of the PR and leave a top-level comment.
+  if(hasProjectOwnerLabel === true) {
+    var labelSubstrings = projectOwnerLabel.split('@');
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: Make sure to deal correctly with the case where the project owner labels are not formatted correctly (multiple '@', no '@", project owner not found in list of People for oppia, etc.)

File: lib/checkPullRequestLabels.js (hunk 1 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: I think we tried this when implementing the checks. It's safe to use `===` for conditionals. We use it throughout the code base.

File: lib/checkPullRequestLabels.js (hunk 4 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: The label should always be in this particular format: `PROJECT: <Project-Name> -- @<Owner-Name>` as defined in the [PR Review Cycle doc](https://docs.google.com/document/d/1y2_3Ns0V00QESZy3gswc413Rki-m1exTNDt3uVcu7w0/edit#).

File: lib/checkPullRequestLabels.js (hunk 3 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Okay, I'll look into this. We can redirect them to the gitter channel if they are unable to decide the project label for a PR.

Comment (apb7) [reply lib/checkPullRequestLabels.js]: I have already discussed with Sandeep and we've decided to close the pull request so that all open PRs always follow the review cycle and the process remains streamlined.

File: lib/checkPullRequestLabels.js (hunk 1 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: Yep, I agree it's safe, but when the variable is a boolean and you're comparing it to a boolean it's fine to use `if(X)` of `if(!X)` for simplicity.

File: lib/checkPullRequestLabels.js (hunk 3 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: OK, sounds good, happy to go with what you think best. Maybe document the design decisions in the PR description.

File: lib/checkPullRequestLabels.js (hunk 4 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: That is the intended format, but mistakes happen. The code should be robust to mistakes and not assume that the world around it is perfectly ordered.

File: lib/checkPullRequestLabels.js (hunk 2 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

File: lib/checkPullRequestLabels.js (hunk 3 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Does it look better now? Thanks!

File: lib/checkPullRequestLabels.js (hunk 1 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: Also please add space after "if"

File: lib/checkPullRequestLabels.js (hunk 3 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: "since it **does** have a project label..." ??

Also, it seems to me like you should make this requirement clear in the PR checkbox. Then, in this error message, you can say something like "as mentioned in the PR checkbox list" to draw attention to that list the next time.

File: lib/checkPullRequestLabels.js (hunk 4 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Sorry, but we will be modifying all the project labels according to this format for the bot to retrieve the project owner's names. There is no other way to get the name of the project owner, therefore, I am sure how to make this more robust.

File: lib/checkPullRequestLabels.js (hunk 3 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: I have modified the corresponding issue description. I'll go ahead and add it here as well. Thanks!

File: lib/checkPullRequestLabels.js (hunk 4 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: No, what I'm trying to say is -- let's say there's a typo in the project owner. Or there's an extra space after the "--". Or someone just goes ahead and adds a label. Or a project owner leaves the codebase.

I'm not saying you need to somehow handle all these cases one by one, but what I am suggesting is that if the "happy path" doesn't work you have some sort of fallback. Basically, what should happen if an error occurs with this automatic process?

Comment (apb7) [reply lib/checkPullRequestLabels.js]: Okay, for extra spaces in the names, the process will work as-is since we're trimming the spaces. For any of the other scenarios, the process will never crash -- it will work but if suppose there's a typo, the project owner name will be incorrect and this will be reflected in the comment, and can, therefore, be corrected by just modifying the label. Or if the project owner leaves, the process will still work and can again be rectified by the comment left by the bot.
Basically, the comment created by the bot will always tell us what's going wrong with the labels.

File: lib/checkPullRequestLabels.js (hunk 3 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks! Also, I've added a TODO regarding listing this instruction in the PR checkbox. Thanks!

File: lib/checkPullRequestLabels.js (hunk 1 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

File: lib/checkPullRequestLabels.js (hunk 4 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: Let's say someone else adds a new project owner label that has no '@' at all, and then adds that label to their PR. What happens then?

Basically, I am trying to suggest the idea of a try/except to catch unforeseen cases.

Comment (apb7) [reply lib/checkPullRequestLabels.js]: For this case, no one will be assigned to the PR and comment will be created which has no project owner. This will make it clear that the label is incorrectly formatted. I can add a try/except for this case where the project owner is an empty string if you'd like. Thanks!

Comment (seanlip) [reply lib/checkPullRequestLabels.js]: Here's what I suggest:

- On obtaining the label, validate that it has exactly the correct format. If it does not, then have the comment explicitly alert the dev workflow team that a project owner cannot be assigned and that there may be a problem with Oppiabot, so that they can look into it.
- After validating the label has the correct format, you can parse it in the usual way, secure in the assurance that the format is good and that there will be no surprises.

Sounds good?

Comment (apb7) [reply lib/checkPullRequestLabels.js]: I have added a regexp test for validation. PTAL, thanks!

File: lib/checkPullRequestLabels.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
+module.exports.checkProjectLabel = async function(context) {
+  var pullRequest = context.payload.pull_request;
+  var pullRequestNumber = pullRequest.number;
+
+  // eslint-disable-next-line no-console
+  console.log('RUNNING PROJECT LABEL CHECK ON PULL REQUEST ' + pullRequestNumber + ' ...');
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: 80-char limit?

File: lib/checkPullRequestLabels.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      var projectOwner = labelSubstrings[labelSubstrings.length - 1].trim();
+
+      var assigneeParams = context.issue({assignees: [projectOwner]});
+      await context.github.issues.addAssigneesToIssue(assigneeParams);
+
+      var commentParams = context.issue({body: 'Assigning @' + projectOwner +
+        ' for the first-pass review of this pull request. Thanks!'});
+      await context.github.issues.createComment(commentParams);
+    }
+    else {
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: Move to end of previous line.

**Comment by apb7** [comment]
Hi @seanlip, ptal, thanks!

File: lib/checkPullRequestLabels.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+  // If the PR has a project label, assign the project owner
+  // for the first-pass review of the PR and leave a top-level comment.
+  if (hasProjectLabel) {
+    // Check if project label is valid.
+
+    // Please refer to https://regex101.com/r/IRWeSA/1 for explanation
+    // of the below regular expression.
+    var regExpForProjectLabel = /^([pP][rR][oO][jJ][eE][cC][tT].*@\s*\w+\s*)/;
+    var isProjectLabelValid = regExpForProjectLabel.test(projectLabel);
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: var projectLabelIsValid

File: lib/checkPullRequestLabels.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    'label, please feel free to ask on ' + gitterLink + '. Thanks!'});
+  await context.github.issues.createComment(stateCommentParams);
+}
+
+module.exports.checkChangelogLabel = async function(context) {
+  var pullRequest = context.payload.pull_request;
+  var pullRequestNumber = pullRequest.number;
+
+  // eslint-disable-next-line no-console
+  console.log('RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST ' + pullRequestNumber + ' ...');
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: 80-char limit?

File: lib/checkPullRequestLabels.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+  var labels = pullRequest.labels;
+  var userName = pullRequest.user.login;
+  var hasChangelogLabel = labels.some(function(label) {
+    return label.name.trim().toUpperCase().startsWith('CHANGELOG');
+  });
+
+  // If the PR has a changelog label, no action is required.
+  if(hasChangelogLabel) return;
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]:
```
if (hasChangelogLabel) {
  return;
}
```

File: lib/checkPullRequestLabels.js (hunk 5 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

File: lib/checkPullRequestLabels.js (hunk 7 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

File: lib/checkPullRequestLabels.js (hunk 6 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

File: lib/checkPullRequestLabels.js (hunk 8 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

File: lib/checkPullRequestLabels.js (hunk 9 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: Done, thanks!

**Comment by apb7** [comment]
> Before doing the full integration, though, please make sure to come to a clear decision on the intended user flow for a new contributor. You might also want to get the relevant labels and PR instructions updated on oppia/oppia first.

Sure, I'll work on this now. Thanks!

**Comment by apb7** [comment]
Hi @seanlip, Sandeep and I have finalized [this workflow](https://github.com/oppia/oppiabot/pull/37#issue-301031650). Please take a look and let us know if it sounds good. Thanks!

**Comment by seanlip** [comment]
Looks good to me. So, project labelling is sort-of optional as far as oppiabot is concerned?

**Comment by seanlip** [comment]
(If so, I'm fine with that, and in general I'm fine with what you and Sandeep agree on. Feel free to merge once you're sure that this won't break the workflow for either new or existing contributors.)

**Comment by apb7** [comment]
Yes, project labelling is optional as a check. We suggest it as an option
to the author when he opens a PR and has signed the CLA. Thanks!

**Comment by apb7** [comment]
Hi Sean, I've made some changes to the code according to the decided workflow. PTAL, if you are around. I've tested these on my local instance and it works fine. Once merged, we can deploy and monitor the actions of the bot. Thanks!

File: index.js [hunk 10]
```diff
@@ -21,6 +22,19 @@ module.exports = (robot) => {
     // the whitelisted accounts.
     if (whitelistedAccounts.includes(context.repo().owner.toLowerCase())) {
       await apiForSheetsModule.checkClaStatus(context);
+      await checkPullRequestLabelsModule.checkChangelogLabel(context);
+    }
+  });
+
+  robot.on('pull_request.reopened', async context => {
+    if (whitelistedAccounts.includes(context.repo().owner.toLowerCase())) {
+      await checkPullRequestLabelsModule.checkChangelogLabel(context);
```
Comment (seanlip) [inline index.js]: Just one thought. Did you check that new contributors (who aren't members of the Oppia repo) are actually able to add labels to PRs?

File: lib/apiForSheets.js [hunk 11]
```diff
@@ -70,8 +70,15 @@ module.exports.generateOutput = async function(rows, prNumber, prDetails) {
     }
 
     // If the PR does not have a CLA label and the author has signed
-    // the CLA, no action is required.
+    // the CLA, welcome the author.
     if (hasUserSignedCla === true) {
+      var commentParams = context.repo({
+        number: prNumber,
+        body: 'Hi! @' + userName + '. Welcome to Oppia! ' +
```
Comment (seanlip) [inline lib/apiForSheets.js]: I don't think you need to thank them for signing the CLA every time they make a PR ...?

File: lib/checkPullRequestLabels.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  var pullRequestNumber = pullRequest.number;
+
+  // eslint-disable-next-line no-console
+  console.log('RUNNING PROJECT LABEL CHECK ON PULL REQUEST ' +
+    pullRequestNumber + ' ...');
+
+  var projectLabel = context.payload.label.name.trim();
+  var hasProjectLabel = projectLabel.toUpperCase().startsWith('PROJECT');
+
+  // If the PR has a project label, assign the project owner
```
Comment (seanlip) [inline lib/checkPullRequestLabels.js]: What if the owner is already assigned?

Also, how do you ensure this comment only comes up once in the PR lifecycle? E.g. the project owner looks at it and approves and then deassigns themselves ... do they get reassigned again?

File: index.js (hunk 10 above)
Comment (apb7) [reply index.js]: New contributors won't be able to add labels until they are on-board (as collaborators). Sandeep and I discussed this -- ideally, their onboarding mentors should help in putting up the labels.

File: lib/apiForSheets.js (hunk 11 above)
Comment (apb7) [reply lib/apiForSheets.js]: Should we add this (to put up a project label when a review is required) in the checklist?

File: lib/checkPullRequestLabels.js (hunk 12 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: This check will run only when a project label is added to a PR by the author, that is, this is triggered by `pull_request.labelled` hook. So, it will basically run once in the PR lifecycle when the author adds a project label.

File: index.js (hunk 10 above)
Comment (seanlip) [reply index.js]: But then wouldn't their PRs immediately be closed as soon as they are opened?

File: lib/apiForSheets.js (hunk 11 above)
Comment (seanlip) [reply lib/apiForSheets.js]: Yes, I think so.

File: index.js (hunk 10 above)
Comment (DubeySandeep) [reply index.js]: Let's not close the PR if the creator isn't a collaborator? Add "New-contributor" label or "Need to assign Changelog" label or add some message to indicate this? (not sure what to do if not closing the PR)

Comment (DubeySandeep) [reply index.js]: In the future, we can assign it to the mentor and add a message asking the mentor to add the label.

Comment (seanlip) [reply index.js]: (Might need to add Project labels to the existing PRs, too)

File: lib/checkPullRequestLabels.js (hunk 12 above)
Comment (seanlip) [reply lib/checkPullRequestLabels.js]: OK.

What happens if there are two project labels added to a PR?

File: index.js (hunk 10 above)
Comment (seanlip) [reply index.js]: If you want to be safe, btw, maybe don't close the PR in general first? Just leave a message asking the author to please fix this?

Then if that's not working we can make this more stringent. But that may be a better incremental step.

If we go this route, maybe file an issue for Oppiabot to ensure that the changelog label is guaranteed to be there at time-of-merge.

Comment (apb7) [reply index.js]: > If you want to be safe, btw, maybe don't close the PR in general first? Just leave a message asking the author to please fix this?
> Then if that's not working we can make this more stringent. But that may be a better incremental step.

I agree. This sounds much better for an incremental change.

> If we go this route, maybe file an issue for Oppiabot to ensure that the changelog label is guaranteed to be there at time-of-merge. (Perhaps it should be implemented as a repo check instead similar to travis-ci etc -- Oppiabot should be more of a helper rather than an enforcer.)

I will do this. Thanks!

File: lib/checkPullRequestLabels.js (hunk 12 above)
Comment (apb7) [reply lib/checkPullRequestLabels.js]: I tested this just now -- the bot will assign all of the project owners according to the label. Each label triggers a separate event on which the bot acts.

**Comment by seanlip** [comment]
OK, thanks @apb7 -- feel free to merge whenever you are ready!

File: index.js (hunk 10 above)
Comment (apb7) [reply index.js]: Filed #38 for this. Thanks!

**Comment by apb7** [comment]
Thanks, @seanlip and @DubeySandeep! Merging this.
